qiskit/synthesis/two_qubit/two_qubit_decompose.py

```diff
@@ -467,6 +217,7 @@ def __repr__(self):
             + b64ascii
             + [
                 f"requested_fidelity={self.requested_fidelity},",
+                f"_specialization={self._inner_decomposition.specialization},"
```

Comment:
the specializations wouldn't typically be in scope so the __repr__ wouldn't normally be eval-able without importing an internal API?

Reply:
I think that's quite normal - they're still importable, and the individual `repr` of any object doesn't usually include qualified access.  Most people don't `from numpy import array`, but `repr(numpy.array([1, 2, 3]))` is `array([1, 2, 3])`, for example.

Reply:
```suggestion
                f"_specialization={self._inner_decomposition.specialization},",
```
(To make the formatting nicer)

Reply:
I'm unsure on this. I agree that it shouldn't  be a requirement that everything can always eval without *any* imports, but making folks figure out that they need to do something pretty non-obvious like `from qiskit._accelerate.two_qubit_decompose import Specialization` in order to use this functionality (which really only exists in the first place to help debugging when things are already confusing). Its unfortunate that the pyo3-generated `__repr__()` for the enum members doesn't include the full module path. Don't block merging based on fixing this, but if you can think of a simple way to make it more user-friendly (without necessarily making the enum part of any public API) that would be good.

Reply:
> Its unfortunate that the pyo3-generated `__repr__()` for the enum members doesn't include the full module path.

What classes are you thinking of where this would be the case?  Python-object reprs aren't fully specified - the `x = eval(repr(x))` thing only ever _really_ applied to Python built-ins anyway, and not even those; there's nothing in `builtins` or `types` that you can import that can make `repr({}.keys())` eval'able, for example, despite it having the "eval-like" form.

Maybe I'm missing something about why the Python-space `__repr__` included a `from_bytes` in the first place, rather than being `f"{self.__class__.__name__}({self.unitary_matrix!r}, requested_fidelity={self.requested_fidelity!r})"` (plus the extra dummy kwargs if you wanted, though I'd argue they're perhaps more relevant for `__str__` than `__repr__`).  It seemed to me that the `from_bytes` method still recalculated all the same things as `__new__` anyway except for having the particular specialisation fixed, and since each class has its own `__new__` determined by `__init_subclass__`, we could have enforced the specialisation through that same mechanism, even in the factory-function `__new__` (i.e., at the point we check `if fidelity is None: instance = object.__new__(cls)`, we could include the same shortcircuit for `if cls is not TwoQubitWeylDecomposition`).

Reply:
Oh, and to answer the last bit about making it more user friendly: honestly, I'm not certain to what degree it's super necessary, but if we were very much concerned about it, we could maybe attach the `Specialization` enum type object as a class attribute on the Python-space type object (and potentially the Rust-space one), and then it'll appear in the documentation of the methods (which people _really_ should read before just copy-paste `eval`ing deep numerical code, I'd think).

Reply:
Honestly I don't completely remember how I ended up with this version of repr. I do remember that it was a brute force rewrite after I originally attempted a couple versions similar to what you suggest and it wasn't enough for getting the consistent debugging I needed for intermittent test fails. Possibly it was the now-fixed numpy array2str bug that bit me. (I believe the subclasses did already enforce the specialization you're suggesting, by forcing fidelity to None in init_subclass.)

Anyway, attaching the enumeration asa class attribute seems reasonable to me. But if I'm the only one who cares about this then I'm ok to be outvoted.

Reply:
I set the enum to a private class attr on the class and used that in the repr so it'll work with `eval()` even without an import: https://github.com/Qiskit/qiskit/pull/11946/commits/4dda4beb6682fac945dfee2d5a6f6a7621a5e076